syntax = "proto2";

package recontaskparameter_pb;

message ReconTaskParameterPB {
    required string task_name = 1;
    required string path_sysmat = 2;
    required string path_sinogdram = 3;
    optional string path_scatter_map = 4;
    optional string path_mu_map = 5;
    required bool use_nn = 6;
    required bool use_scatter_map = 7;
    required int32 num_iters = 8;
    required int32 num_dual_iters = 9;
    required double gamma = 10;
    required double lambda = 11;
    optional double coeff_scatter = 12;
    required string output_dir = 13;
    enum IteratorType {
        MLEM = 0;
        EM_Tikhonov = 1;
        PAPA_2DWavelet = 2;
        PAPA_Cont = 3;
        PAPA_Cont_TV = 4;
        PAPA_Cont_Wavelet = 5;
        PAPA_Dynamic = 6;
        PAPA_TV = 7;
    }
    required IteratorType iterator = 14;
    enum FileFormat {
        RAW_SINOGRAM = 0;
        RAW_PROJECTION = 1;
        DICOM = 2;
    }
    required FileFormat file_format = 15;
    required int32 num_input_images = 16;
    enum FileDataType {
        FLOAT32 = 1;
        FLOAT64 = 2;
        INT8 = 3;
        UINT8 = 4;
        INT16 = 5;
        UINT16 = 6;
        INT32 = 7;
        UINT32 = 8;
        INT64 = 9;
        UINT64 = 10;
    }
    required FileDataType file_data_type = 17;

    required int32 index_sinogram = 18;
    required int32 index_projection = 19;
    required int32 num_detectors = 20;
    required int32 num_angles = 21;
    required int32 num_slices = 22;
    required int32 num_results = 23;

    repeated double sinogram_data = 24; // (num_slices, num_angles, num_detectors)
    repeated double restored_sinogram_data = 25;
    repeated double reconstructed_tomographs_data = 26;

    required bool done = 27; // Whether the task is completed.
    required bool loaded = 28;
    required string path_model = 29; // THe path to the neural network model for scattering correction.

    required string sinogram_info = 30;
    required int32 resolution = 31;
}
